cibuildwheel/options.py

```diff
@@ -150,28 +152,73 @@ class ConfigOptionError(KeyError):
     pass
 
 
-def _dig_first(*pairs: tuple[Mapping[str, Setting], str], ignore_empty: bool = False) -> Setting:
+class InheritRule(enum.Enum):
+    NONE = enum.auto()
+    APPEND = enum.auto()
+    PREPEND = enum.auto()
+
+
+def _resolve_cascade(
+    *pairs: tuple[Setting | None, InheritRule], ignore_empty: bool = False
+) -> Setting:
     """
-    Return the first dict item that matches from pairs of dicts and keys.
-    Will throw a KeyError if missing.
+    Given a cascade of values with inherit rules, resolve them into a single
+    value.
+
+    'None' values mean that the option was not set at that level.
 
-    _dig_first((dict1, "key1"), (dict2, "key2"), ...)
+    Values start with defaults, followed by more specific rules. If rules
+    are NONE, the last non-null value is returned. If a rule is APPEND or
+    PREPEND, the value is concatenated with the previous value, according to
+    the rules in _merge_values.
+
+    The following idiom can be used to get the first matching value:
+
+        _resolve_cascade(("value1", Inherit.NONE), ("value2", Inherit.NONE), ...)))
     """
     if not pairs:
         msg = "pairs cannot be empty"
         raise ValueError(msg)
 
-    for dict_like, key in pairs:
-        if key in dict_like:
-            value = dict_like[key]
+    result: Setting | None = None
 
-            if ignore_empty and value == "":
-                continue
+    for value, rule in pairs:
+        if value is None:
+            continue
 
-            return value
+        if ignore_empty and not value:
+            continue
+
+        if result is None:  # noqa: SIM108
+            result = value
+        else:
+            result = _merge_values(result, value, rule)
+
+    if result is None:
+        msg = "a setting should at least have a default value"
+        raise ValueError(msg)
 
-    last_key = pairs[-1][1]
-    raise KeyError(last_key)
+    return result
+
+
+def _merge_values(before: Setting, after: Setting, rule: InheritRule) -> Setting:
+    if rule == InheritRule.NONE:
+        return after
+
+    if isinstance(before, list) and isinstance(after, list):
+        if rule == InheritRule.APPEND:
+            return before + after
+        else:
+            return after + before
+
+    if isinstance(before, dict) and isinstance(after, dict):
+        if rule == InheritRule.APPEND:
+            return {**before, **after}
+        else:
+            return {**after, **before}
+
+    msg = f"Cannot merge {type(before)} and {type(after)} with {rule}"
+    raise TypeError(msg)
```

Comment:
This is something I'd like to improve. This makes the following config illegal:

```
[tool.cibuildwheel]
environment = "A=1"

[tool.cibuildwheel.overrides]
select = "something"
environment = "B=2"
inherit.environment = "append"
```

I think we should pass in the sep and table params so they can be used to merge the values here.

Reply:
Ahhhh, this is more complicated than I thought! In the case where we have:
```toml
[tool.cibuildwheel]
environment = {A="1" B="2"}

[tool.cibuildwheel.overrides]
select = "something"
environment = {A="override"}
inherit.environment = "append"
```

We'd like to get back `A=override B=2`. But, if we represent the values internally to the cascade as strings, we'd get back `A=1 B=2 A=override`. The behaviour would be the same, however.

But.... actually, now that I think about it, for environment, I think that is actually more correct. Consider this config:

```toml
[tool.cibuildwheel]
environment = {PATH="/opt/bin:$PATH"}

[tool.cibuildwheel.overrides]
select = "something"
environment = {PATH="/opt/local/bin:$PATH"}
inherit.environment = "append"
```

If we were to overwrite the PATH key in the override, we'd lose the entry of `/opt/bin/`. So I think it's actually more useful to simply string-append these settings.